### PR TITLE
Copy slec4py to install

### LINKED ISSUE #3033: INSTALL: We do not copy the slepc4py source directory
**Describe the error**
We updated the install script when pip changed behaviour to copy the petsc4py source, then install so that petsc4py could be installed from a centrally installed PETSc. We should do the same for selpc4py when there is a centrally installed SLEPc.

**Steps to Reproduce**
1. System install SLEPc
2. `firedrake-install --honour-petsc-dir --slepc`

**Expected behavior**
Firedrake with slepc4py installed in venv.

**Error message**
```
Using installed SLEPc from /opt/slepc/firedrake/full-opt/
Installing slepc4py/
Traceback (most recent call last):
  File "/scratch/jbetteri/firedrake_py311_opt/src/../bin/firedrake-update", line 2095, in <module>
    install("slepc4py/")
  File "/scratch/jbetteri/firedrake_py311_opt/src/../bin/firedrake-update", line 1050, in install
    run_pip_install(["--ignore-installed", get_slepc4py_dir()])
  File "/scratch/jbetteri/firedrake_py311_opt/src/../bin/firedrake-update", line 979, in run_pip_install
    check_call(pipinstall + pipargs)
  File "/scratch/jbetteri/firedrake_py311_opt/src/../bin/firedrake-update", line 666, in check_call
    log.debug(subprocess.check_output(arguments, stderr=subprocess.STDOUT, env=os.environ).decode())
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/python/v3.11.2/lib/python3.11/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/python/v3.11.2/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/scratch/jbetteri/firedrake_py311_opt/bin/python', '-m', 'pip', 'install', '--no-build-isolation', '--no-binary', 'mpi4py,randomgen,islpy,numpy', '-
-no-deps', '-vvv', '--ignore-installed', '/opt/slepc/firedrake/full-opt/src/binding/slepc4py']' returned non-zero exit status 1.


Update log saved in /scratch/jbetteri/firedrake_py311_opt/firedrake-update.log
```

```
...
Building wheels for collected packages: slepc4py
  Created temporary directory: /tmp/pip-wheel-smoc1dvu
  Building wheel for slepc4py (setup.py): started
  Destination directory: /tmp/pip-wheel-smoc1dvu
  Running command python setup.py bdist_wheel
  running bdist_wheel
  running build
  running build_src
  running build_py
  error: [Errno 1] Operation not permitted: 'build/lib.linux-x86_64-cpython-311/slepc4py/SLEPc.pxd'
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /scratch/jbetteri/firedrake_py311_opt/bin/python -u -c '
  exec(compile('"'"''"'"''"'"'
  # This is <pip-setuptools-caller> -- a caller that pip uses to run setup.py
  #
  # - It imports setuptools before invoking setup.py, to enable projects that directly
  #   import from `distutils.core` to work with newer packaging standards.
  # - It provides a clear error message when setuptools is not installed.
  # - It sets `sys.argv[0]` to the underlying `setup.py`, when invoking `setup.py` so
  #   setuptools doesn'"'"'t think the script is `-c`. This avoids the following warning:
  #     manifest_maker: standard file '"'"'-c'"'"' not found".
  # - It generates a shim setup.py, for handling setup.cfg-only projects.
  import os, sys, tokenize
  
  try:
      import setuptools
  except ImportError as error:
      print(
          "ERROR: Can not execute `setup.py` since setuptools is not available in "
          "the build environment.",
          file=sys.stderr,
      )
      sys.exit(1)
  
  __file__ = %r
  sys.argv[0] = __file__
  
  if os.path.exists(__file__):
      filename = __file__
      with tokenize.open(__file__) as f:
          setup_py_code = f.read()
  else:
      filename = "<auto-generated setuptools caller>"
      setup_py_code = "from setuptools import setup; setup()"
  
  exec(compile(setup_py_code, filename, "exec"))
  '"'"''"'"''"'"' % ('"'"'/opt/slepc/firedrake/full-opt/src/binding/slepc4py/setup.py'"'"',), "<pip-setuptools-caller>", "exec"))' bdist_wheel -d /tmp/pip-wheel-smoc1dvu
  cwd: /opt/slepc/firedrake/full-opt/src/binding/slepc4py/
  Building wheel for slepc4py (setup.py): finished with status 'error'
  ERROR: Failed building wheel for slepc4py
  Running setup.py clean for slepc4py
  Running command python setup.py clean
  running clean
  removing 'build/temp.linux-x86_64-cpython-311' (and everything under it)
  error removing build/temp.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/temp.linux-x86_64-cpython-311/full-debug/src/slepc4py/SLEPc.o'
  error removing build/temp.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/temp.linux-x86_64-cpython-311/full-debug/src/slepc4py'
  error removing build/temp.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/temp.linux-x86_64-cpython-311/full-debug/src'
  error removing build/temp.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/temp.linux-x86_64-cpython-311/full-debug'
  error removing build/temp.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/temp.linux-x86_64-cpython-311/full-opt/src/slepc4py/SLEPc.o'
  error removing build/temp.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/temp.linux-x86_64-cpython-311/full-opt/src/slepc4py'
  error removing build/temp.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/temp.linux-x86_64-cpython-311/full-opt/src'
  error removing build/temp.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/temp.linux-x86_64-cpython-311/full-opt'
  error removing build/temp.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/temp.linux-x86_64-cpython-311/complex-opt/src/slepc4py/SLEPc.o'
  error removing build/temp.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/temp.linux-x86_64-cpython-311/complex-opt/src/slepc4py'
  error removing build/temp.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/temp.linux-x86_64-cpython-311/complex-opt/src'
  error removing build/temp.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/temp.linux-x86_64-cpython-311/complex-opt'
  error removing build/temp.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/temp.linux-x86_64-cpython-311'
  removing 'build/lib.linux-x86_64-cpython-311' (and everything under it)
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/lib/slepc.cfg'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/lib/full-debug/SLEPc.cpython-311-x86_64-linux-gnu.so'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/lib/full-debug'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/lib/full-opt/SLEPc.cpython-311-x86_64-linux-gnu.so'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/lib/full-opt'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/lib/__init__.py'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/lib/complex-opt/SLEPc.cpython-311-x86_64-linux-gnu.so'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/lib/complex-opt'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/lib'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/SLEPc.pxd'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/include/slepc4py/slepc4py.h'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/include/slepc4py/slepc4py.i'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/include/slepc4py'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/include'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/SLEPc_api.h'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/SLEPc.py'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/__main__.py'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/__init__.py'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py/SLEPc.h'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311/slepc4py'
  error removing build/lib.linux-x86_64-cpython-311: [Errno 13] Permission denied: 'build/lib.linux-x86_64-cpython-311'
  removing 'build/bdist.linux-x86_64' (and everything under it)
  error removing build/bdist.linux-x86_64: [Errno 13] Permission denied: 'build/bdist.linux-x86_64'
  'build/scripts-3.11' does not exist -- can't clean it
Failed to build slepc4py
ERROR: Could not build wheels for slepc4py, which is required to install pyproject.toml-based projects
Exception information:
Traceback (most recent call last):
  File "/scratch/jbetteri/firedrake_py311_opt/lib/python3.11/site-packages/pip/_internal/cli/base_command.py", line 180, in exc_logging_wrapper
    status = run_func(*args)
             ^^^^^^^^^^^^^^^
  File "/scratch/jbetteri/firedrake_py311_opt/lib/python3.11/site-packages/pip/_internal/cli/req_command.py", line 248, in wrapper
    return func(self, options, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/jbetteri/firedrake_py311_opt/lib/python3.11/site-packages/pip/_internal/commands/install.py", line 429, in run
    raise InstallationError(
pip._internal.exceptions.InstallationError: Could not build wheels for slepc4py, which is required to install pyproject.toml-based projects
Remote version of pip: 23.2
Local version of pip:  23.2
Was pip installed by pip? True
Removed build tracker: '/tmp/pip-build-tracker-9dyqpun0'

2023-07-19 13:15:49,345 INFO   

Update log saved in /scratch/jbetteri/firedrake_py311_opt/firedrake-update.log
```

**Environment:**
 - OS: Any
 - Python version: Any
 - PETSC_DIR and SLEPC_DIR set

### PR DESCRIPTION
**Pull request by JDBetteridge**
# Description
Make `build_and_install_petsc4py()` install script function generic to handle slepc4py.

## Associated Pull Requests:
- N/A

## Fixes Issues:
- Fixes #3033 

# Checklist for author:

- [ ] I have linted the codebase (`make lint` in the `firedrake` source directory).
- [ ] My changes generate no new warnings.
- [ ] All of my functions and classes have appropriate docstrings.
- [ ] I have commented my code where its purpose may be unclear.
- [ ] I have included and updated any relevant documentation.
- [ ] Documentation builds locally (`make linkcheck; make html; make latexpdf` in `firedrake/docs` directory)
- [ ] I have added tests specific to the issues fixed in this PR.
- [ ] I have added tests that exercise the new functionality I have introduced
- [ ] Tests pass locally (`pytest tests` in the `firedrake` source directory) (useful, but not essential if you don't have suitable hardware).
- [ ] I have performed a self-review of my own code using the below guidelines.

# Checklist for reviewer:

- [ ] Docstrings present
- [ ] New tests present
- [ ] Code correctly commented
- [ ] No bad "code smells"
- [ ] No issues in parallel
- [ ] No CI issues (excessive parallelism/memory usage/time/warnings generated)
- [ ] Upstream/dependent branches and PRs are ready

### CODE REVIEW

**Comment by JDBetteridge** [comment]
Tested working on the Pileus workstation

File: scripts/firedrake-install [hunk 1]
```diff
@@ -1165,26 +1164,32 @@ def build_and_install_petsc():
                 check_call(["rm", "-f", deletefile])
 
 
-def build_and_install_petsc4py():
+def build_and_install_x4py(package="petsc4py"):
+    if package == "petsc4py":
+        get_x4py_dir = get_petsc4py_dir
+    elif package == "slepc4py":
+        get_x4py_dir = get_slepc4py_dir
```
Comment (dham) [inline scripts/firedrake-install:1171]:
```suggestion
def build_and_install_x4py(package="petsc4py"):
    if package == "petsc4py":
        x4py_dir = get_petsc4py_dir()
    elif package == "slepc4py":
        x4py_dir = get_slepc4py_dir()
```

File: scripts/firedrake-install [hunk 2]
```diff
@@ -1165,26 +1164,32 @@ def build_and_install_petsc():
...
         if args.honour_petsc_dir:
-            log.info("Copying petsc4py to virtualenv to install")
-            petsc4py_copy = os.path.join(os.environ['VIRTUAL_ENV'], 'src', 'petsc4py')
-            if not os.path.exists(petsc4py_copy):
-                shutil.copytree(get_petsc4py_dir(), petsc4py_copy, dirs_exist_ok=True)
-            pipargs = ["-vvv", "--ignore-installed", petsc4py_copy]
+            log.info("Copying {} to virtualenv to install".format(package))
+            x4py_copy = os.path.join(os.environ['VIRTUAL_ENV'], 'src', package)
+            if not os.path.exists(x4py_copy):
+                shutil.copytree(get_x4py_dir(), x4py_copy, dirs_exist_ok=True)
```
Comment (dham) [inline scripts/firedrake-install:1179]:
```suggestion
                shutil.copytree(x4py_dir, x4py_copy, dirs_exist_ok=True)
```

File: scripts/firedrake-install [hunk 3]
```diff
@@ -1165,26 +1164,32 @@ def build_and_install_petsc():
...
+                    fh.write("# Unable to install {}\n".format(package))
+                    fh.write("This is just a copy of the {} source for debugging\n".format(package))
                     fh.write("See `firedrake-install.log` for the error")
                 raise
             else:
-                shutil.rmtree(petsc4py_copy)
+                shutil.rmtree(x4py_copy)
         else:
-            pipargs = ["-vvv", "--ignore-installed", get_petsc4py_dir()]
+            pipargs = ["-vvv", "--ignore-installed", get_x4py_dir()]
```
Comment (dham) [inline scripts/firedrake-install:1192]:
```suggestion
            pipargs = ["-vvv", "--ignore-installed", x4py_dir]
```

**Review by dham — CHANGES_REQUESTED** [review]
Too clever.

**Review by dham — APPROVED** [review]

**Pull request merged**